crates/ruff_cli/src/cache.rs

```diff
@@ -1,171 +1,240 @@
+    }
 
-            s.serialize_element(&SerializeMessage { message, file_id })?;
+    /// Get the cached results for a single file at relative `path`. This uses
+    /// `file_last_modified` to determine if the results are still accurate
+    /// (i.e. if the file hasn't been modified since the cached run).
+    ///
+    /// This returns `None` if `file_last_modified` differs from the cached
+    /// timestamp or if the cache doesn't contain results for the file.
+    pub(crate) fn get(
+        &self,
+        path: &RelativePath,
+        file_last_modified: SystemTime,
+    ) -> Option<FileCache> {
+        let files = self.files.lock().unwrap();
+        let file = files.get(path)?;
+
+        // Make sure the file hasn't changed since the cached run.
+        if file.last_modified != file_last_modified {
+            return None;
         }
 
-        s.end()
+        Some(file.clone())
+    }
+
+    /// Add or update a file cache at `path` relative to the package root.
+    pub(crate) fn update(&self, path: RelativePathBuf, file: FileCache) {
+        self.files.lock().unwrap().insert(path, file);
     }
-}
 
-struct SerializeMessage<'a> {
-    message: &'a Message,
-    file_id: usize,
+    /// Remove a file cache at `path` relative to the package root.
+    pub(crate) fn remove(&self, path: &RelativePath) {
+        self.files.lock().unwrap().remove(path);
+    }
 }
 
-impl Serialize for SerializeMessage<'_> {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let Message {
-            kind,
-            range,
-            fix,
-            // Serialized manually for all files
-            file: _,
-            noqa_offset: noqa_row,
-        } = self.message;
+/// [`Path`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePath = Path;
+/// [`PathBuf`] that is relative to the package root in [`PackageCache`].
+pub(crate) type RelativePathBuf = PathBuf;
 
-        let mut s = serializer.serialize_struct("Message", 5)?;
+/// On disk representation of the cache per source file.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub(crate) struct FileCache {
+    /// Timestamp when the file was last modified before the (cached) check.
+    last_modified: SystemTime,
+    /// Imports made.
+    imports: ImportMap,
+    /// Diagnostic messages.
+    messages: Vec<CacheMessage>,
+    /// Source code of the file.
+    ///
+    /// # Notes
+    ///
+    /// This will be empty if `messages` is empty.
+    source: String,
+}
+
+impl FileCache {
+    /// Create a new source file cache.
+    pub(crate) fn new(
+        last_modified: SystemTime,
+        messages: &[Message],
+        imports: &ImportMap,
+    ) -> FileCache {
+        let source = if let Some(msg) = messages.first() {
+            msg.file.source_text().to_owned()
+        } else {
+            String::new() // No messages, no need to keep the source!
+        };
```

Comment:
We should add at least an assertion (release build) that asserts that the source files of all messages are identical.

Reply:
> We should add at least an assertion (release build) that asserts that the source files of all messages are identical.

Can add :+1: 

> I don't know if there are any diagnostics today that make use of the multi-file diagnostic source frames (I assume not) @charliermarsh ?

Last week @charliermarsh said not yet